File: Assets/Script/Lib/Agent/AgentRuntimePolicy.lua
-- [ts]: AgentRuntimePolicy.ts
local ____lualib = require("lualib_bundle") -- 1
local __TS__StringSplit = ____lualib.__TS__StringSplit -- 1
local __TS__StringStartsWith = ____lualib.__TS__StringStartsWith -- 1
local __TS__StringTrim = ____lualib.__TS__StringTrim -- 1
local __TS__ArraySlice = ____lualib.__TS__ArraySlice -- 1
local __TS__ObjectAssign = ____lualib.__TS__ObjectAssign -- 1
local ____exports = {} -- 1
local ____Dora = require("Dora") -- 2
local Content = ____Dora.Content -- 2
local Path = ____Dora.Path -- 2
local ____Utils = require("Agent.Utils") -- 4
local sanitizeUTF8 = ____Utils.sanitizeUTF8 -- 4
local ____WebIDESync = require("Agent.Tool.WebIDESync") -- 5
local sendWebIDEFileUpdate = ____WebIDESync.sendWebIDEFileUpdate -- 5
local Tools = require("Agent.Tools") -- 6
____exports.AGENT_PLAN_DIR = ".agent/plan" -- 8
____exports.AGENT_PLAN_FILE = ".agent/plan/PLAN.md" -- 9
____exports.AGENT_PROGRESS_FILE = ".agent/plan/PROGRESS.md" -- 10
local DEFAULT_PLAN_DOCUMENT = "# 开发方案\n\n## 目标\n\n## 背景与当前实现\n\n## 范围\n\n### 包含\n\n### 不包含\n\n## 已确认决策\n\n## 待确认问题\n\n无\n\n## 技术方案\n\n## 实施步骤\n\n| ID | 工作项 | 依赖 | 验收条件 |\n| --- | --- | --- | --- |\n\n## 风险与回退方案\n\n## 验证计划\n\n## 变更记录\n"
local DEFAULT_PROGRESS_DOCUMENT = "# 开发进度\n\n## 当前工作\n\n## 步骤进度\n\n| ID | 状态 | 最新结果 | 下一步 |\n| --- | --- | --- | --- |\n\n## 修改记录\n\n## 验证证据\n\n## 阻塞问题\n\n## 进度日志\n"
local function trimText(value) -- 62
	local trimmed = string.match(value, "^%s*(.-)%s*$") -- 63
	return trimmed or "" -- 64
end -- 62
function ____exports.normalizeAgentPath(path) -- 67
	local normalized = table.concat( -- 68
		__TS__StringSplit( -- 68
			trimText(path), -- 68
			"\\" -- 68
		), -- 68
		"/" -- 68
	) -- 68
	while __TS__StringStartsWith(normalized, "./") do -- 68
		normalized = string.sub(normalized, 3) -- 69
	end -- 69
	return normalized -- 70
end -- 67
function ____exports.getAgentDecisionPath(input) -- 73
	if type(input.path) == "string" then -- 73
		return __TS__StringTrim(input.path) -- 74
	end -- 74
	if type(input.target_file) == "string" then -- 74
		return __TS__StringTrim(input.target_file) -- 75
	end -- 75
	return "" -- 76
end -- 73
function ____exports.isMainAgentMemoryPath(path) -- 79
	local normalized = ____exports.normalizeAgentPath(path) -- 80
	return normalized == ".agent/main" or __TS__StringStartsWith(normalized, ".agent/main/") -- 81
end -- 79
function ____exports.isAgentPlanPath(path) -- 84
	local normalized = ____exports.normalizeAgentPath(path) -- 85
	return normalized == ____exports.AGENT_PLAN_DIR or __TS__StringStartsWith(normalized, ____exports.AGENT_PLAN_DIR .. "/") -- 86
end -- 84
function ____exports.isAgentInternalDocumentPath(path) -- 89
	return ____exports.isMainAgentMemoryPath(path) or ____exports.isAgentPlanPath(path) -- 90
end -- 89
local function ensureDirectory(dir) -- 93
	if Content:exist(dir) then -- 93
		return Content:isdir(dir) -- 94
	end -- 94
	local parent = Path:getPath(dir) -- 95
	if parent ~= "" and parent ~= dir and not Content:exist(parent) and not ensureDirectory(parent) then -- 95
		return false -- 96
	end -- 96
	return Content:mkdir(dir) -- 97
end -- 93
function ____exports.ensureAgentPlanDocuments(workDir) -- 100
	local dir = Path(workDir, ____exports.AGENT_PLAN_DIR) -- 101
	if not ensureDirectory(dir) then -- 101
		return {success = false, message = "failed to create " .. ____exports.AGENT_PLAN_DIR} -- 102
	end -- 102
	local created = {} -- 103
	local documents = {{____exports.AGENT_PLAN_FILE, DEFAULT_PLAN_DOCUMENT}, {____exports.AGENT_PROGRESS_FILE, DEFAULT_PROGRESS_DOCUMENT}} -- 104
	do -- 104
		local i = 0 -- 108
		while i < #documents do -- 108
			do -- 108
				local relative, content = table.unpack(documents[i + 1], 1, 2) -- 109
				local path = Path(workDir, relative) -- 110
				if Content:exist(path) then -- 110
					goto __continue17 -- 111
				end -- 111
				if not Content:save(path, content) then -- 111
					return {success = false, message = "failed to create " .. relative} -- 112
				end -- 112
				sendWebIDEFileUpdate(path, true, content) -- 113
				created[#created + 1] = relative -- 114
			end -- 114
			::__continue17:: -- 114
			i = i + 1 -- 108
		end -- 108
	end -- 108
	return {success = true, created = created} -- 116
end -- 100
function ____exports.isEditBudgetExhausted(state) -- 127
	local mustCreateFreshEntry = state.freshProjectBuildPending == true and state.freshProjectCodeFile == nil and state.hasBuilt ~= true -- 128
	return state.unbuiltEdits == true and (state.editsSinceBuild or 0) >= 3 and not mustCreateFreshEntry -- 131
end -- 127
function ____exports.getUncoveredConversationMessages(messages, lastConsolidatedIndex) -- 136
	return __TS__ArraySlice(messages, lastConsolidatedIndex) -- 137
end -- 136
function ____exports.normalizeLineEndings(text) -- 140
	return table.concat( -- 141
		__TS__StringSplit( -- 141
			table.concat( -- 141
				__TS__StringSplit(text, "\r\n"), -- 141
				"\n" -- 141
			), -- 141
			"\r" -- 141
		), -- 141
		"\n" -- 141
	) -- 141
end -- 140
function ____exports.countOccurrences(text, needle) -- 144
	if needle == "" then -- 144
		return 0 -- 145
	end -- 145
	local count = 0 -- 146
	local start = 0 -- 147
	while start <= #text - #needle do -- 147
		local index = (string.find( -- 149
			text, -- 149
			needle, -- 149
			math.max(start + 1, 1), -- 149
			true -- 149
		) or 0) - 1 -- 149
		if index < 0 then -- 149
			break -- 150
		end -- 150
		count = count + 1 -- 151
		start = index + #needle -- 152
	end -- 152
	return count -- 154
end -- 144
function ____exports.containsWholeFileDuplicate(existing, replacement) -- 157
	local normalizedExisting = ____exports.normalizeLineEndings(existing) -- 158
	local normalizedReplacement = ____exports.normalizeLineEndings(replacement) -- 159
	if #normalizedExisting < 16 or #normalizedReplacement <= #normalizedExisting then -- 159
		return false -- 160
	end -- 160
	return ____exports.countOccurrences(normalizedReplacement, normalizedExisting) > 1 -- 161
end -- 157
function ____exports.successfulEditResult(workDir, path, base) -- 164
	local current = Tools.readFileRaw(workDir, path) -- 169
	local currentCharacters = current.success and type(current.content) == "string" and #current.content or 0 -- 170
	return __TS__ObjectAssign( -- 171
		{}, -- 171
		base, -- 172
		{ -- 171
			actualSaved = current.success, -- 173
			actualSavedCharacters = currentCharacters, -- 174
			currentFileExists = current.success, -- 175
			currentCharacters = currentCharacters, -- 176
			currentState = current.success and (("saved " .. tostring(currentCharacters)) .. " characters to ") .. path or "file state unavailable after edit: " .. sanitizeUTF8(current.message) -- 177
		} -- 177
	) -- 177
end -- 164
return ____exports -- 164

File: Assets/Script/Lib/Agent/AgentRuntimePolicy.ts
// @preview-file off clear
import { Content, Path } from "Dora";
import type { Message } from "Agent/Utils";
import { estimateTextTokens, safeJsonEncode, sanitizeUTF8 } from "Agent/Utils";
import * as Tools from "Agent/Tools";

export const AGENT_PLAN_DIR = ".agent/plan";
export const AGENT_PLAN_FILE = ".agent/plan/PLAN.md";
export const AGENT_PROGRESS_FILE = ".agent/plan/PROGRESS.md";

const DEFAULT_PLAN_DOCUMENT = `# 开发方案

## 目标

## 背景与当前实现

## 范围

### 包含

### 不包含

## 已确认决策

## 待确认问题

无

## 技术方案

## 实施步骤

| ID | 工作项 | 依赖 | 验收条件 |
| --- | --- | --- | --- |

## 风险与回退方案

## 验证计划

## 变更记录
`;

const DEFAULT_PROGRESS_DOCUMENT = `# 开发进度

## 当前工作

## 步骤进度

| ID | 状态 | 最新结果 | 下一步 |
| --- | --- | --- | --- |

## 修改记录

## 验证证据

## 阻塞问题

## 进度日志
`;

function trimText(value: string): string {
	const [trimmed] = string.match(value, "^%s*(.-)%s*$");
	return trimmed ?? "";
}

export function normalizeAgentPath(path: string): string {
	let normalized = trimText(path).split("\\").join("/");
	while (normalized.startsWith("./")) normalized = normalized.slice(2);
	return normalized;
}

export function isMainAgentMemoryPath(path: string): boolean {
	const normalized = normalizeAgentPath(path);
	return normalized === ".agent/main" || normalized.startsWith(".agent/main/");
}

export function isAgentPlanPath(path: string): boolean {
	const normalized = normalizeAgentPath(path);
	return normalized === AGENT_PLAN_DIR || normalized.startsWith(`${AGENT_PLAN_DIR}/`);
}

export function isAgentInternalDocumentPath(path: string): boolean {
	return isMainAgentMemoryPath(path) || isAgentPlanPath(path);
}

function ensureDirectory(dir: string): boolean {
	if (Content.exist(dir)) return Content.isdir(dir);
	const parent = Path.getPath(dir);
	if (parent !== "" && parent !== dir && !Content.exist(parent) && !ensureDirectory(parent)) return false;
	return Content.mkdir(dir);
}

export function ensureAgentPlanDocuments(workDir: string): { success: true; created: string[] } | { success: false; message: string } {
	const dir = Path(workDir, AGENT_PLAN_DIR);
	if (!ensureDirectory(dir)) return { success: false, message: `failed to create ${AGENT_PLAN_DIR}` };
	const created: string[] = [];
	const documents: Array<[string, string]> = [
		[AGENT_PLAN_FILE, DEFAULT_PLAN_DOCUMENT],
		[AGENT_PROGRESS_FILE, DEFAULT_PROGRESS_DOCUMENT],
	];
	for (let i = 0; i < documents.length; i++) {
		const [relative, content] = documents[i];
		const path = Path(workDir, relative);
		if (Content.exist(path)) continue;
		if (!Content.save(path, content)) return { success: false, message: `failed to create ${relative}` };
		Tools.sendWebIDEFileUpdate(path, true, content);
		created.push(relative);
	}
	return { success: true, created };
}

export interface EditBudgetState {
	freshProjectBuildPending?: boolean;
	freshProjectCodeFile?: string;
	hasBuilt?: boolean;
	unbuiltEdits?: boolean;
	editsSinceBuild?: number;
}

export function isEditBudgetExhausted(state: EditBudgetState): boolean {
	const mustCreateFreshEntry = state.freshProjectBuildPending === true
		&& state.freshProjectCodeFile === undefined
		&& state.hasBuilt !== true;
	return state.unbuiltEdits === true
		&& (state.editsSinceBuild ?? 0) >= 3
		&& !mustCreateFreshEntry;
}

export function getUncoveredConversationMessages(messages: Message[], lastConsolidatedIndex: number): Message[] {
	return messages.slice(lastConsolidatedIndex);
}

export function estimateConversationTokens(messages: Message[]): number {
	let tokens = 0;
	for (let i = 0; i < messages.length; i++) {
		const message = messages[i];
		tokens += 8;
		tokens += estimateTextTokens(message.role ?? "");
		tokens += estimateTextTokens(message.content ?? "");
		tokens += estimateTextTokens(message.name ?? "");
		tokens += estimateTextTokens(message.tool_call_id ?? "");
		tokens += estimateTextTokens(message.reasoning_content ?? "");
		const [toolCallsText] = safeJsonEncode((message.tool_calls ?? []) as object);
		tokens += estimateTextTokens(toolCallsText ?? "");
	}
	return tokens;
}

export function normalizeLineEndings(text: string): string {
	return text.split("\r\n").join("\n").split("\r").join("\n");
}

export function countOccurrences(text: string, needle: string): number {
	if (needle === "") return 0;
	let count = 0;
	let start = 0;
	while (start <= text.length - needle.length) {
		const index = text.indexOf(needle, start);
		if (index < 0) break;
		count++;
		start = index + needle.length;
	}
	return count;
}

export function containsWholeFileDuplicate(existing: string, replacement: string): boolean {
	const normalizedExisting = normalizeLineEndings(existing);
	const normalizedReplacement = normalizeLineEndings(replacement);
	if (normalizedExisting.length < 16 || normalizedReplacement.length <= normalizedExisting.length) return false;
	return countOccurrences(normalizedReplacement, normalizedExisting) > 1;
}

export function successfulEditResult(
	workDir: string,
	path: string,
	base: Record<string, unknown>
): Record<string, unknown> {
	const current = Tools.readFileRaw(workDir, path);
	const currentCharacters = current.success && typeof current.content === "string" ? current.content.length : 0;
	return {
		...base,
		actualSaved: current.success,
		actualSavedCharacters: currentCharacters,
		currentFileExists: current.success,
		currentCharacters,
		currentState: current.success
			? `saved ${currentCharacters} characters to ${path}`
			: `file state unavailable after edit: ${sanitizeUTF8(current.message)}`,
	};
}
